### ISSUE TITLE
OpenCL errors on Ubuntu with latest drivers

### ISSUE THREAD

**Issue by glinscott** [glinscott/leela-chess#124]
I'm seeing errors like this occasionally, which take my worker down:
```
Error in OpenCL calculation: expected -0.001008 got -0.001075 (271(error=6.699571%)
```

The go client is currently coded to exit on errors like this...  So, a few possibilities:
1. The error threshold needs to be increased/changed
2. Have the client log the error, but continue
3. Figure out if this is a reproducible failure

**Comment by Error323** [comment]
I'm experiencing the same error, on my client I ripped out the exception. I thought I was affected because of having two gpus, which makes it worse. I suggest we rip out the exception and just log the error? So option 2.

I did some experiments and it only happens occasionally, but the current thresholds are too strict. Also I didn't like that code I wrote there anyway. I'll remove it OK?

**Comment by jkiliani** [comment]
What about changing the self-check code so it automatically logs the CPU and GPU NN outputs into a file? At the moment it looks quite possible that the error threshold is a bit too strict, maybe the more complicated heads in lczero compared to leelaz require an adjusted threshold?

**Comment by Error323** [comment]
Yes I think that's correct. The errors are higher because our heads are have much bigger matrix multiplications.

**Comment by killerducky** [comment]
More logging? On the error call opencl again and see if the result changes? I'm nervous because I've never had this error on my machine. How often does it happen?

**Comment by killerducky** [comment]
If it's not a bug, shouldn't you be able to record what position causes the error, and reproduce the error 100% of the time? If we can reproduce it then we could do some analysis to see if it's expected or not.

**Comment by Error323** [comment]
What is your hardware @killerducky ? I think it only occurs with NVIDIA graphics cards.

**Comment by Error323** [comment]
I've got access to 2 GTX 1080 and 2 GTX 1080Ti's and all 4 experience the problem, tried various drivers.

**Comment by killerducky** [comment]
I have an older nvidia card: Ubuntu | i5-7600K 3.80GHz | GTX 970

I'm only running one client, and I have two LZGo bots running on CGOS, no ponder, so they are much lighter load.

**Comment by Error323** [comment]
Yeah I also got a 560Ti that's working without any issues. I don't know what we should do here. Shall we just rip out the error and start logging some metadata? I gotta fix #119 first though, and the README...

**Comment by killerducky** [comment]
If you don't get to it first I can add tests/logging tonight or tomorrow. I can fake some errors to test it since my machine isn't getting the errors. I think short term it's ok since it happens rarely it will just be noise in the training.

**Comment by Error323** [comment]
Awesome!

**Comment by jjoshua2** [comment]
I have 1080ti on windows 10 x64 that has never had an error for leela go or chess. I've updated drivers 2 or 3 times as well, so it wasn't just one magic driver.

**Comment by killerducky** [comment]
See pull #135

**Comment by killerducky** [comment]
logosgg is getting this error, he posted debug logs. Looks like his client has the -debug option, but his lzchess does not have the OpenCL debug logs. I asked him to download from https://ci.appveyor.com/project/glinscott/leela-chess/branch/next so maybe we can get more info.

The log has a little private information in it (his Windows user name, some info on his hardware) so I'm asking his permission to post it. Here is just the error message:

```
info depth 15 nodes 271 nps 572 score cp -2 winrate 49.92% time 473 pv e2e4 e7e5 d2d4 d7d6 c2c3 g8f6
info depth 16 nodes 482 nps 630 score cp 6 winrate 50.25% time 764 pv e2e4 e7e5 d2d4 d7d6 d4d5 g8f6
Error in OpenCL calculation: expected 1.338263 got 2.771106 (503(error=107.067418%)
Error in OpenCL calculation: expected 0.012664 got 1.088928 (3(error=8498.880768%)
Update your GPU drivers or reduce the amount of games played simultaneously.
```

**Comment by jkiliani** [comment]
Did anyone yet do some strength testing of CPU versus GPU clients, at same net and playouts? That should clear up whether the OpenCL errors are of a magnitude that's cause for concern...
